[IRGen] Add ability to disable compact value witnesses from block list

drexin opened the pull request:
rdar://124629183

Allows usage of compact value witnesses to be disabled using a block list.

drexin commented:
@swift-ci smoke test

tshortli commented on lib/IRGen/GenValueWitness.cpp:
```diff
@@ -885,10 +886,20 @@ void addStride(ConstantStructBuilder &B, const TypeInfo *TI, IRGenModule &IGM) {
 
+bool irgen::layoutStringsEnabled(IRGenModule &IGM) {
+  auto moduleName = IGM.getSwiftModule()->getRealName();
+  if (IGM.Context.blockListConfig.hasBlockListAction(moduleName.str(), BlockListKeyKind::ModuleName, BlockListAction::ShouldUseLayoutStringValueWitnesses)) {
```
Nit: `clang-format`

tshortli replied:
You may also want to emit a diagnostic once somewhere indicating that the block list has taken effect to ensure that there is a record if when someone is confused about why it is disabled.

drexin committed:
[IRGen] Add ability to disable compact value witnesses from block list

rdar://124629183

Allows usage of compact value witnesses to be disabled using a block list.

drexin commented:
@swift-ci smoke test

tshortli commented on lib/IRGen/IRGen.cpp:
```diff
@@ -1162,6 +1163,12 @@ GeneratedModule IRGenRequest::evaluate(Evaluator &evaluator,
 
+  if (!layoutStringsEnabled(IGM)) {
+    auto moduleName = IGM.getSwiftModule()->getRealName().str();
+    IGM.Context.Diags.diagnose(SourceLoc(), diag::layout_strings_blocked,
```
Won't this diagnose for normal programs that don't enable CVWT, even if they are not on the block list?

drexin commented:
@swift-ci smoke test

drexin commented:
@swift-ci smoke test windows
